### PR TITLE
WIP main:  factor the top-level as textToYAML

### PR DESCRIPTION
**Pull request by deepfire**
@Gabriel439, this is a suggested refactoring -- would you feel like including something like this?

If so, what improvements (besides the obvious lack of a docstring) would you suggest?

### CODE REVIEW

**Comment by Gabriel439** [comment]
Perhaps simplify the utility to only encompass this logic:

```haskell
codeToValue :: Text -> Text -> IO Value
codeToValue name code = do
    expr <- case Dhall.Parser.exprFromText (Directed name 0 0 0 0) code of
              Left  err  -> Control.Exception.throwIO err
              Right expr -> return expr

    expr' <- Dhall.Import.load expr
    case Dhall.TypeCheck.typeOf expr' of
      Left  err -> Control.Exception.throwIO err
      Right _   -> return ()

    case Dhall.JSON.dhallToJSON expr' of
      Left  err  -> Control.Exception.throwIO err
      Right json -> return json
```

Then if the user wants detailed error messages they can wrap that in `detailed` and if they want to omit null then they wrap the result in `omitNull`.  Also, returning `Value` has the advantage that you can generate both YAML and JSON from the same utility.

**Comment by deepfire** [comment]
@Gabriel439, does that one look better?

File: src/Dhall/YAML.hs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+{-# LANGUAGE DeriveDataTypeable #-}
```
Comment (Gabriel439) [inline src/Dhall/YAML.hs]: I don't think you need this module any longer.  All it does is re-export `codeToValue` from the `Dhall.JSON` module

File: dhall-to-yaml/Main.hs [hunk 2]
```diff
@@ -31,31 +30,15 @@ data Options = Options
...
-        json <- case Dhall.JSON.dhallToJSON expr' of
-            Left err  -> Control.Exception.throwIO err
-            Right json -> return json
-
-        let filteredJSON =
-                if Options.Generic.unHelpful omitNull
-                then Dhall.JSON.omitNull json
-                else json
-        Data.ByteString.putStr (Data.Yaml.encode filteredJSON) ))
+    Data.ByteString.putStr =<< (Data.Yaml.encode . omittingNull <$>) . explaining . (Dhall.JSON.codeToValue "(stdin)")
```
Comment (Gabriel439) [inline dhall-to-yaml/Main.hs]: I'd prefer to use `do` notation here to split this up over multiple lines.  This is difficult to read

File: src/Dhall/YAML.hs (hunk 1 above)
Comment (deepfire) [reply src/Dhall/YAML.hs]: Right, removed it.

File: dhall-to-yaml/Main.hs (hunk 2 above)
Comment (deepfire) [reply dhall-to-yaml/Main.hs]: Done!

File: dhall-json.cabal [hunk 3]
```diff
@@ -30,10 +30,14 @@ Library
     Hs-Source-Dirs: src
     Build-Depends:
         base                 >= 4.8.0.0  && < 5   ,
+        bytestring                          < 0.11,
         aeson                >= 1.0.0.0  && < 1.4 ,
         dhall                >= 1.11.0   && < 1.12,
         text                 >= 0.11.1.0 && < 1.3 ,
-        unordered-containers                < 0.3
+        trifecta             >= 1.6      && < 1.8 ,
```
Comment (Gabriel439) [inline dhall-json.cabal]: I think you can now remove the `trifecta` dependency from the `dhall-to-json` and `dhall-to-yaml` executables

**Comment by deepfire** [comment]
Added a provisional docstring to `textToValue`.

File: src/Dhall/JSON.hs [hunk 4]
```diff
@@ -192,3 +198,27 @@ omitNull (Bool bool) =
...
+
+>>> :set -XOverloadedStrings
+>>> :set -XOverloadedLists
+>>> import Dhall.Core
+>>> Dhall.JSON.codeToValue "(stdin)" "{ a = 1 }"
+>>> Object (fromList [("a",Number 1.0)])
+-}
+codeToValue :: Data.Text.Text -> Data.Text.Text -> IO Value
+codeToValue name code = do
+    let textBS = Data.ByteString.UTF8.fromString . Data.Text.unpack
```
Comment (Gabriel439) [inline src/Dhall/JSON.hs]: Oops!  Perhaps I should have suggested that the first argument have type `ByteString` that way we don't need to convert from `Text` using the `utf8-string` package

File: dhall-json.cabal [hunk 5]
```diff
@@ -30,10 +30,14 @@ Library
...
+        unordered-containers                < 0.3 ,
+        utf8-string                               ,
+        yaml                 >= 0.5.0    && < 0.9
```
Comment (Gabriel439) [inline dhall-json.cabal]: I think the library section doesn't need to depend on the `yaml` package any longer now that `codeToValue` returns a `Value`

File: src/Dhall/JSON.hs [hunk 6]
```diff
@@ -192,3 +198,27 @@ omitNull (Bool bool) =
     Bool bool
 omitNull Null =
     Null
+
+{-| Convert a piece of Text carrying a Dhall inscription to an equivalent JSON Value
+
+>>> :set -XOverloadedStrings
+>>> :set -XOverloadedLists
```
Comment (Gabriel439) [inline src/Dhall/JSON.hs]: I think this example doesn't require `-XOverloadedLists`

File: src/Dhall/JSON.hs [hunk 7]
```diff
@@ -192,3 +198,27 @@ omitNull (Bool bool) =
...
+>>> import Dhall.Core
+>>> Dhall.JSON.codeToValue "(stdin)" "{ a = 1 }"
+>>> Object (fromList [("a",Number 1.0)])
+-}
+codeToValue :: Data.Text.Text -> Data.Text.Text -> IO Value
```
Comment (Gabriel439) [inline src/Dhall/JSON.hs]: Don't forget to document what each argument represents

File: src/Dhall/JSON.hs (hunk 4 above)
Comment (deepfire) [reply src/Dhall/JSON.hs]: Done.

File: dhall-json.cabal (hunk 3 above)
Comment (deepfire) [reply dhall-json.cabal]: Done.

File: dhall-json.cabal (hunk 5 above)
Comment (deepfire) [reply dhall-json.cabal]: Done.

File: src/Dhall/JSON.hs (hunk 6 above)
Comment (deepfire) [reply src/Dhall/JSON.hs]: Done.

**Comment by deepfire** [comment]
All done and CI passes : -)

**Comment by Gabriel439** [comment]
Looks great!  Thanks for doing this :)

**Comment by deepfire** [comment]
Thank you for your feedback!

**Comment by Gabriel439** [comment]
You're welcome! :)

**Pull request merged**